src/common/src/array/stream_chunk.rs

```diff
@@ -306,6 +307,102 @@ impl EstimateSize for StreamChunk {
     }
 }
 
+enum OpsMutState {
+    ArcRef(Arc<[Op]>),
+    MutRef(Vec<Op>),
+    Undefined,
+}
+
+pub struct OpsMut {
+    state: OpsMutState,
+}
+
+impl OpsMut {
+    pub fn new(ops: Arc<[Op]>) -> Self {
+        Self {
+            state: OpsMutState::ArcRef(ops),
+        }
+    }
+
+    pub fn set(&mut self, n: usize, val: Op) {
+        if let OpsMutState::MutRef(v) = &mut self.state {
+            v[n] = val;
+        } else {
+            let state = mem::replace(&mut self.state, OpsMutState::Undefined); // intermediate state
+            let mut v = match state {
+                OpsMutState::ArcRef(v) => v.to_vec(),
+                OpsMutState::MutRef(_) | OpsMutState::Undefined => unreachable!(),
+            };
+            v[n] = val;
+            self.state = OpsMutState::MutRef(v);
+        }
+    }
+}
+impl From<OpsMut> for Arc<[Op]> {
+    fn from(v: OpsMut) -> Self {
+        match v.state {
+            OpsMutState::ArcRef(a) => a,
+            OpsMutState::MutRef(v) => v.into(),
+            OpsMutState::Undefined => unreachable!(),
+        }
+    }
+}
+
+struct StreamChunkMut {
+    columns: Arc<[ArrayRef]>,
+    ops: OpsMut,
+    vis: VisMut,
+}
+
+impl From<StreamChunk> for StreamChunkMut {
+    fn from(c: StreamChunk) -> Self {
+        let (c, ops) = c.into_parts();
+        let (columns, vis) = c.into_parts_v2();
+        Self {
+            columns,
+            ops: OpsMut::new(ops),
+            vis: vis.into(),
+        }
+    }
+}
+
+impl From<StreamChunkMut> for StreamChunk {
+    fn from(c: StreamChunkMut) -> Self {
+        StreamChunk::from_parts(c.ops, DataChunk::from_parts(c.columns, c.vis.into()))
+    }
+}
+pub struct OpRowMutRef<'a> {
+    c: &'a mut StreamChunkMut,
+    i: usize,
+}
+
+impl OpRowMutRef<'_> {
+    pub fn set_vis(&mut self, val: bool) {
+        self.c.set_vis(self.i, val);
+    }
+
+    pub fn set_op(&mut self, val: Op) {
+        self.c.set_op(self.i, val);
+    }
+}
+
+impl StreamChunkMut {
+    pub fn set_vis(&mut self, n: usize, val: bool) {
+        self.vis.set(n, val);
+    }
+
+    pub fn set_op(&mut self, n: usize, val: Op) {
+        self.ops.set(n, val);
+    }
+
+    pub unsafe fn to_mut_rows(&self) -> impl Iterator<Item = OpRowMutRef<'_>> {
```

Comment:
I feel this function can be safe?
```suggestion
    pub fn iter_rows_mut(&mut self) -> impl Iterator<Item = OpRowMutRef<'_>> {
```

Reply:
BTW, can we get rid of `unsafe` by manually implementing the iterator?

Reply:
:thinking:

Reply:
> I guess it's safe only if we bound the lifetime of returning iterator.

The iterator's items are borrowing `self` so `+ '_` is not needed. 

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3f253acc4590bfd3de4a619be201c65b

Reply:
> BTW, can we get rid of unsafe by manually implementing the iterator?

Each item has a `&mut` to the chunk and multiple items can exist at the same time. So I guess it's impossible to implement in safe Rust.

Reply:
fixed.